### PR TITLE
US89668 - adding support for additive schema.

### PR DESCRIPTION
**Pull request by neverendingqs**
Interested in thoughts on complexity of the solution.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -97,6 +98,42 @@ def get_plugin_link_mapping(config, access_token):
...
+        csv_rows.append(line[:num_columns_in_table])
+
+    csv_data = io.StringIO()
+    csv_writer = csv.writer(csv_data, quoting=csv.QUOTE_MINIMAL)
+    csv_writer.writerows(csv_rows)
+
+    # Reset the cursor so that readers read start reading from the beginning of
+    # the stream, not the end
+    csv_data.seek(io.SEEK_SET)
+    return csv_data
```
Comment (neverendingqs) [inline main.py]: Parsing the CSV in Python (vs. letting PostgreSQL do it) will have to look something like this regardless if we do streams or not. The added complexity here seems unavoidable.

File: main.py [hunk 2]
```diff
@@ -97,6 +98,41 @@ def get_plugin_link_mapping(config, access_token):
...
+    for line in csv_reader:
+        csv_rows.append(line[:num_columns_in_table])
+
+    csv_data = io.StringIO()
+    csv_writer = csv.writer(csv_data, quoting=csv.QUOTE_MINIMAL)
+    csv_writer.writerows(csv_rows)
+
+    # Rewind the stream to the beginning before returning it
+    csv_data.seek(io.SEEK_SET)
+    return csv_data
```
Comment (neverendingqs) [inline main.py]: Parsing the CSV in Python (vs. letting PostgreSQL do it) will have to look something like this regardless if we do streams or not. The added complexity here seems unavoidable.

File: main.py [hunk 3]
```diff
@@ -97,6 +98,41 @@ def get_plugin_link_mapping(config, access_token):
...
+
+def process_csv_stream(csv_input_stream, num_columns_in_table):
+    '''
+    Ignore excessive columns in the CSV due to additive changes / BDS minor
+    changes by ignoring any columns in the CSV past the number of columns in the
+    table
+    '''
+    csv_rows = []
+    csv_reader = csv.reader(csv_input_stream, quoting=csv.QUOTE_MINIMAL)
+    for line in csv_reader:
```
Comment (lboyd) [inline main.py]: Do we ever express caveats with customer examples: for example this code may not be performant for very large CSV files, etc.

File: main.py [hunk 4]
```diff
@@ -97,6 +98,41 @@ def get_plugin_link_mapping(config, access_token):
...
+        csv_rows.append(line[:num_columns_in_table])
+
+    csv_data = io.StringIO()
+    csv_writer = csv.writer(csv_data, quoting=csv.QUOTE_MINIMAL)
+    csv_writer.writerows(csv_rows)
+
+    # Rewind the stream to the beginning before returning it
+    csv_data.seek(io.SEEK_SET)
```
Comment (lboyd) [inline main.py]: Potentially unnecessary as we don't guarantee sort order in the CSV file - probably not a big deal just thought you should know :)

File: main.py [hunk 5]
```diff
@@ -145,6 +181,39 @@ def update_db(db_conn_params, table, csv_data):
 
         conn.commit()
 
+def batch_update_db(db_conn_params, table, csv_file, batch_size=10000):
```
Comment (lboyd) [inline main.py]: Great job with breaking this into consumable chunks!

**Comment by neverendingqs** [comment]
Forgot to add - after figuring out how to work with streams (didn't realize I had to rewind the stream), it was easy to batch it as well, so I added it to this example to not introduce a OOM issue.

Caveat: neither before or after was tested with large data sets.

**Comment by neverendingqs** [comment]
Rebased onto tip of `master`.

**Comment by neverendingqs** [comment]
Not going to merge until I had a chance to properly tag the other code sample.

**Pull request merged**